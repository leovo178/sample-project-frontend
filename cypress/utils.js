/* eslint-disable no-undef */
const generateRandomNumber = (min, max) => {
	min = Math.ceil(min)
	max = Math.floor(max)
	return Math.floor(Math.random() * (max - min)) + min
}

const clickItemByTestId = testId =>
	cy
		.findByTestId(testId)
		.should('exist')
		.click()

const mockMessage = 'mockmessage'

export { generateRandomNumber, clickItemByTestId, mockMessage }
