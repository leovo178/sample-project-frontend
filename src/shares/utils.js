const AUTH_TOKEN = 'AUTH_TOKEN'

const getToken = () => localStorage.getItem(AUTH_TOKEN)
const setToken = token => localStorage.setItem(AUTH_TOKEN, token)
const deleteToken = () => localStorage.removeItem(AUTH_TOKEN)

const validateEmail = email => {
	const emailRegex = /^[a-z][a-z0-9_\.]{4,32}@[a-z0-9]{2,}(\.[a-z0-9]{2,4}){1,2}$/gm
	return email.match(emailRegex)
}

const validateName = name => {
	const nameRegex = /^[a-zA-Z0-9]{3,30}$/
	return name.match(nameRegex)
}

const validatePassword = password => {
	const passwordRegex = /^[a-zA-Z0-9]{3,}$/
	return password.match(passwordRegex)
}

export {
	getToken,
	setToken,
	deleteToken,
	validateEmail,
	validateName,
	validatePassword,
}
