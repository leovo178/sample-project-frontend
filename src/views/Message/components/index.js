export { default as CreateMessageBox } from './CreateMessageBox'

export { default as SearchMessageBox } from './SearchMessageBox'
