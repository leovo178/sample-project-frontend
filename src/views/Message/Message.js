import React, { useState } from 'react'

import { useMutation, useQuery } from '@apollo/react-hooks'

import { Box, Grid, makeStyles } from '@material-ui/core'

import {
	Loading,
	LargeTable,
	DeleteDialog,
	ModifyDialog,
} from '@views_components'
import { CreateMessageBox, SearchMessageBox } from './components'

import {
	MESSAGE_LIST,
	GET_MESSAGE_SEARCH_TEXT,
	SET_MESSAGE_SEARCH_TEXT,
	SET_MESSAGE_CREATE_TEXT,
} from './gql/query'
import { CREATE_MESSAGE, DELETE_MESSAGE, UPDATE_MESSAGE } from './gql/mutation'
import { useCreateMessage, useDeleteMessage } from './gql/useMutations'

import { NETWORK_STATUS_FETCH_MORE, PAGE_LIMIT } from '@src/configs.local'

const useStyle = makeStyles(theme => ({
	root: {
		width: 'calc(100% - 80px)',
		height: '100vh',
		padding: theme.spacing(3),
		position: 'relative',
	},
	container: {
		border: `1px solid ${theme.palette.common.border}`,
		height: '100%',
	},
	item__actionbox: {
		padding: theme.spacing(2),
	},
	divider: {
		borderTop: `1px solid ${theme.palette.common.border}`,
	},
}))

const tableHeaders = [
	{ headerLabel: 'DATE', xs: 5, headerVariable: 'lastModified' },
	{ headerLabel: 'CONTENT', xs: 7, headerVariable: 'content' },
]

const Message = () => {
	const classes = useStyle()

	const [deleteDialogVisible, setDeleteDialogVisible] = useState(false)
	const [confirmDialogVisible, setConfirmDialogVisible] = useState(false)

	const [selectedMessage, setSelectedMessage] = useState(false)

	const {
		data: { messageSearchValueOfMessage },
	} = useQuery(GET_MESSAGE_SEARCH_TEXT)

	const { loading, error, data, fetchMore, networkStatus } = useQuery(
		MESSAGE_LIST,
		{
			variables: { query: { limit: PAGE_LIMIT } },
			notifyOnNetworkStatusChange: true,
			onError: err => alert(err),
		}
	)

	const [setMessageSearchValueOfMain] = useMutation(SET_MESSAGE_SEARCH_TEXT)
	const [setMessageCreateValueOfMain] = useMutation(SET_MESSAGE_CREATE_TEXT)

	const [createMessage] = useCreateMessage(CREATE_MESSAGE, MESSAGE_LIST, {
		query: { limit: PAGE_LIMIT },
	})

	const [deleteMessage] = useDeleteMessage(DELETE_MESSAGE, MESSAGE_LIST, {
		query: { limit: PAGE_LIMIT },
	})

	const [updateMessage] = useMutation(UPDATE_MESSAGE, {
		onError: err => alert(err),
	})

	const handleCreateMessage = createVal => {
		createMessage({
			variables: { message: { content: createVal } },
		}).then(() =>
			setMessageCreateValueOfMain({ variables: { createValue: '' } })
		)
	}

	const handleSearch = value => {
		setMessageSearchValueOfMain({ variables: { searchValue: value } })
		if (value === messageSearchValueOfMessage) {
			return false
		} else {
			try {
				fetchMore({
					variables: {
						query: {
							limit: PAGE_LIMIT,
							searchText: value,
						},
					},
					updateQuery: (prev, { fetchMoreResult }) => {
						if (!fetchMoreResult) return prev
						const fetchedMessageList = fetchMoreResult.messageList
						let cacheMessageList = prev.messageList
						const hasNext = fetchedMessageList.hasNext

						return {
							messageList: {
								...cacheMessageList,
								items: fetchedMessageList.items,
								hasNext,
							},
						}
					},
				})
			} catch (error) {
				alert(error.message)
			}
		}
	}

	const loadNextMessagePage = () => {
		try {
			fetchMore({
				variables: {
					query: {
						limit: PAGE_LIMIT,
						skip: data.messageList.items.length,
						searchText: messageSearchValueOfMessage,
					},
				},
				updateQuery: (prev, { fetchMoreResult }) => {
					if (!fetchMoreResult) return prev
					const fetchedMessageList = fetchMoreResult.messageList
					let cacheMessageList = prev.messageList

					const items = [...cacheMessageList.items, ...fetchedMessageList.items]
					const hasNext = fetchedMessageList.hasNext

					return {
						messageList: {
							...cacheMessageList,
							items,
							hasNext,
						},
					}
				},
			})
		} catch (error) {
			alert(error.message)
		}
	}

	const handleOnSelectMessage = message => {
		setConfirmDialogVisible(true)
		setSelectedMessage(message)
	}

	const handleOnDeleteMessage = message => {
		setDeleteDialogVisible(true)
		setSelectedMessage(message)
	}

	const handleOnAgreeDeleteMessage = () => {
		setDeleteDialogVisible(false)
		deleteMessage({ variables: { id: selectedMessage.id } })
	}

	const handleOnAgreeModifyMessage = message => {
		updateMessage({
			variables: { message: { id: selectedMessage.id, content: message } },
		})
	}

	if (error) return <p>Error :(</p>

	return (
		<Box className={classes.root}>
			<Grid container direction='column' className={classes.container}>
				<Grid item className={classes.item__actionbox}>
					<CreateMessageBox
						width={328}
						placeholder='text...'
						onSubmit={handleCreateMessage}
					/>
				</Grid>
				<div className={classes.divider} />
				<Grid item className={classes.item__actionbox}>
					<SearchMessageBox
						width={328}
						placeholder='search...'
						type='search'
						defaultValue={messageSearchValueOfMessage}
						onSubmit={handleSearch}
					/>
				</Grid>

				{loading && networkStatus !== NETWORK_STATUS_FETCH_MORE ? (
					<Loading open={true} msg={'Loading...'} />
				) : (
					<LargeTable
						items={data.messageList.items}
						onClickRow={handleOnSelectMessage}
						handleDeleteRow={handleOnDeleteMessage}
						selectedRow={selectedMessage}
						columns={tableHeaders}
						loadingMore={networkStatus === NETWORK_STATUS_FETCH_MORE}
						isIconClose={true}
						loadNextPage={loadNextMessagePage}
						hasNextPage={data.messageList.hasNext}
					/>
				)}
			</Grid>

			<DeleteDialog
				open={deleteDialogVisible}
				onClose={() => {
					setDeleteDialogVisible(false)
				}}
				onAgree={handleOnAgreeDeleteMessage}
				onDisagree={() => {
					setDeleteDialogVisible(false)
				}}
			/>
			<ModifyDialog
				open={confirmDialogVisible}
				onClose={() => {
					setConfirmDialogVisible(false)
				}}
				valueDefault={selectedMessage.content}
				onAgree={handleOnAgreeModifyMessage}
				onDisagree={() => {
					setConfirmDialogVisible(false)
				}}
			/>
		</Box>
	)
}

export default Message
