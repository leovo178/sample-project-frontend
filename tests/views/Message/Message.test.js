import React from 'react'

import waait from 'waait'

import { act, fireEvent, waitFor } from '@testing-library/react'

import Message from '@views/Message'

import {
	openModifyDialog,
	openDeleteDialog,
	modifyMessage,
	closeDialog,
} from '@tests/shares/utils'
import { findDOMNodeOfMessage, mockSearchText, mockMessage } from './utils'

describe('<Message />', () => {
	let rendered

	beforeEach(async () => {
		await act(async () => {
			rendered = findDOMNodeOfMessage(<Message />)
		})
	})

	it('should match snapshot', async () => {
		await waait(10)

		expect(rendered).toMatchSnapshot()
	})

	it('should fetch successfully new message list with a search text value', async () => {
		const input = rendered.querySelectorAll('[placeholder="search..."]')[0]
		const searchButton = rendered.querySelectorAll(
			'[data-testid=search-icon]'
		)[0]

		expect(input.value).toBe('')

		await act(async () => {
			fireEvent.change(input, { target: { value: mockSearchText } })
		})

		expect(input.value).toBe(mockSearchText)

		await act(async () => {
			fireEvent.click(searchButton)
		})

		expect(rendered.textContent).toContain(mockSearchText)
	})

	it('should allow creating a valid message', async () => {
		const input = rendered.querySelectorAll('[placeholder="text..."]')[0]
		const saveButton = rendered.querySelectorAll(
			'[data-testid=actioninputbox-button]'
		)[0]

		expect(input.value).toBe('')

		await act(async () => {
			fireEvent.change(input, { target: { value: mockMessage } })
		})

		expect(input.value).toBe(mockMessage)

		await act(async () => {
			fireEvent.click(saveButton)
		})

		expect(rendered.textContent).toContain(mockMessage)
	})

	it('should not allow creating an empty message', async () => {
		const saveButton = rendered.querySelectorAll(
			'[data-testid=actioninputbox-button]'
		)[0]

		jest
			.spyOn(window, 'alert')
			.mockImplementation(
				() =>
					'Error: GraphQL error: "message.content" is not allowed to be empty'
			)

		await act(async () => {
			fireEvent.click(saveButton)
		})

		expect(window.alert).toHaveBeenCalled()
	})

	describe('ModifyDialog', () => {
		it('should allow to click No button to close dialog', async () => {
			await openModifyDialog(rendered)

			const disagreeButton = document.querySelectorAll(
				`[data-testid=modifydialog-disagreebutton]`
			)[0]

			await act(async () => {
				await disagreeButton.click()
			})

			waitFor(() => {
				expect(
					document.querySelectorAll(`[data-testid=modifydialog-title]`)[0]
				).toBeFalsy()
			})
		})

		it('should allow to modify message', async () => {
			await waait(10)
			await openModifyDialog(rendered)
			await modifyMessage(mockMessage)

			waitFor(() => {
				expect(
					document.querySelectorAll(`[data-testid=modifydialog-title]`)[0]
				).toBeFalsy()
			})
		})

		it('should not allow to submit a empty message', async () => {
			jest
				.spyOn(window, 'alert')
				.mockImplementation(
					() =>
						'Error: GraphQL error: "message.content" is not allowed to be empty'
				)
			await waait(10)
			await openModifyDialog(rendered)
			await modifyMessage('')

			expect(window.alert).toHaveBeenCalled()

			waitFor(() => {
				expect(
					document.querySelectorAll(`[data-testid=modifydialog-title]`)[0]
				).toBeFalsy()
			})
		})

		it('should allow to click MuiBackdrop-root to close ModifyDialog', async () => {
			await waait(10)
			await openModifyDialog(rendered)
			await closeDialog()

			waitFor(() => {
				expect(
					document.querySelectorAll(`[data-testid=modifydialog-title]`)[0]
				).toBeFalsy()
			})
		})
	})

	describe('DeleteDialog', () => {
		it('should allow to click Yes button', async () => {
			await waait(10)
			await openDeleteDialog(rendered)

			const agreeButton = document.querySelectorAll(
				`[data-testid=deletedialog-agreebutton]`
			)[0]

			await act(async () => {
				agreeButton.click()
			})

			waitFor(() => {
				expect(
					document.querySelectorAll(`[data-testid=deletedialog-title]`)[0]
				).toBeFalsy()
			})
		})

		it('should allow to click No button', async () => {
			await waait(10)
			await openDeleteDialog(rendered)

			const disagreeButton = document.querySelectorAll(
				`[data-testid=deletedialog-disagreebutton]`
			)[0]

			await act(async () => {
				disagreeButton.click()
			})

			waitFor(() => {
				expect(
					document.querySelectorAll(`[data-testid=deletedialog-title]`)[0]
				).toBeFalsy()
			})
		})

		it('should allow to click MuiBackdrop-root to close DeleteDialog', async () => {
			await waait(10)
			await openDeleteDialog(rendered)
			await closeDialog()

			waitFor(() => {
				expect(
					document.querySelectorAll(`[data-testid=deletedialog-title]`)[0]
				).toBeFalsy()
			})
		})
	})
})
