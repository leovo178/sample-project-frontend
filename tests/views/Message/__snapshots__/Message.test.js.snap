// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Message /> should match snapshot 1`] = `
<div
  style="width: 300px; height: 700px;"
>
  <div
    class="MuiBox-root MuiBox-root-6 makeStyles-root-1"
  >
    <div
      class="MuiGrid-root makeStyles-container-2 MuiGrid-container MuiGrid-direction-xs-column"
    >
      <div
        class="MuiGrid-root makeStyles-item__actionbox-3 MuiGrid-item"
      >
        <div
          class="MuiBox-root MuiBox-root-115 makeStyles-root-110"
        >
          <div
            class="MuiFormControl-root MuiTextField-root makeStyles-input-111 makeStyles-input-114"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input"
                placeholder="text..."
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-154 MuiOutlinedInput-notchedOutline"
                style="padding-left: 8px;"
              >
                <legend
                  class="PrivateNotchedOutline-legend-155"
                  style="width: 0.01px;"
                >
                  <span>
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-112 MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge"
            data-testid="actioninputbox-button"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Save
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="makeStyles-divider-4"
      />
      <div
        class="MuiGrid-root makeStyles-item__actionbox-3 MuiGrid-item"
      >
        <div
          class="MuiBox-root MuiBox-root-191 makeStyles-root-110"
        >
          <div
            class="MuiFormControl-root MuiTextField-root makeStyles-input-111 makeStyles-input-190"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input"
                placeholder="search..."
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-154 MuiOutlinedInput-notchedOutline"
                style="padding-left: 8px;"
              >
                <legend
                  class="PrivateNotchedOutline-legend-155"
                  style="width: 0.01px;"
                >
                  <span>
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-112 MuiButton-containedSizeLarge MuiButton-sizeLarge"
            data-testid="actioninputbox-button"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root makeStyles-icon-113"
                data-testid="search-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="MuiBox-root MuiBox-root-238 makeStyles-root-235"
        data-testid="infinitetable"
      >
        <div
          class="MuiGrid-root makeStyles-container__header-236 MuiGrid-container"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
          >
            <p
              class="MuiTypography-root makeStyles-item__header-237 MuiTypography-body2"
              data-testid="tableheader-lastModified"
            >
              DATE
            </p>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7"
          >
            <p
              class="MuiTypography-root makeStyles-item__header-237 MuiTypography-body2"
              data-testid="tableheader-content"
            >
              CONTENT
            </p>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-269 makeStyles-root-235"
        >
          <div
            style="overflow: visible; height: 0px; width: 0px;"
          >
            <div
              class="makeStyles-overflow-270"
              style="height: 700px; width: 200px; overflow: auto; position: relative; will-change: transform;"
            >
              <div
                style="height: 11900px; width: 100%;"
              >
                <div
                  class="MuiGrid-root makeStyles-root-271 makeStyles-root-276 MuiGrid-container"
                  data-testid="row-5e902a09035ac411cd04d22b"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root makeStyles-item__closeicon-273"
                    data-testid="row-closeicon-5e902a09035ac411cd04d22b"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-277 makeStyles-columns__item-274"
                    >
                      <p
                        class="MuiTypography-root makeStyles-columns__item__typo-275 MuiTypography-body2"
                      >
                        2020-04-10T08:10:49+00:00
                      </p>
                    </div>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-278 makeStyles-columns__item-274"
                    >
                      <p
                        class="MuiTypography-root makeStyles-columns__item__typo-275 MuiTypography-body2"
                      >
                        1212
                      </p>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiGrid-root makeStyles-root-271 makeStyles-root-279 MuiGrid-container"
                  data-testid="row-5e6b34a80f14940526bb0780"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root makeStyles-item__closeicon-273"
                    data-testid="row-closeicon-5e6b34a80f14940526bb0780"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-280 makeStyles-columns__item-274"
                    >
                      <p
                        class="MuiTypography-root makeStyles-columns__item__typo-275 MuiTypography-body2"
                      >
                        2020-04-10T08:10:47+00:00
                      </p>
                    </div>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-281 makeStyles-columns__item-274"
                    >
                      <p
                        class="MuiTypography-root makeStyles-columns__item__typo-275 MuiTypography-body2"
                      >
                        13s123x232s sadaa
                      </p>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiGrid-root makeStyles-root-271 makeStyles-root-282 MuiGrid-container"
                  data-testid="row-5e8a886b71b355002f81c196"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root makeStyles-item__closeicon-273"
                    data-testid="row-closeicon-5e8a886b71b355002f81c196"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-283 makeStyles-columns__item-274"
                    >
                      <p
                        class="MuiTypography-root makeStyles-columns__item__typo-275 MuiTypography-body2"
                      >
                        2020-04-06T01:39:55+00:00
                      </p>
                    </div>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-284 makeStyles-columns__item-274"
                    >
                      <p
                        class="MuiTypography-root makeStyles-columns__item__typo-275 MuiTypography-body2"
                      >
                        1212312
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="resize-triggers"
          >
            <div
              class="expand-trigger"
            >
              <div
                style="width: 201px; height: 701px;"
              />
            </div>
            <div
              class="contract-trigger"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
